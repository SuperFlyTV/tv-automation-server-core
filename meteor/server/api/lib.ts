import { Meteor } from 'meteor/meteor'
import { MethodContext } from '../../lib/api/methods'
import { RundownPlaylistId, RundownPlaylist } from '../../lib/collections/RundownPlaylists'
import { Rundown, RundownId } from '../../lib/collections/Rundowns'
import { RundownPlaylistContentAccess, RundownPlaylistContentWriteAccess } from '../security/rundownPlaylist'

/**
 * This is returned from a check of access to a playlist, when access is granted.
 * Fields will be populated about the user.
 * It is identical to RundownPlaylistContentAccess, except for confirming access is allowed
 */
export interface VerifiedRundownPlaylistContentAccess extends RundownPlaylistContentAccess {
	playlist: RundownPlaylist
}

export function checkAccessToPlaylist(
	context: MethodContext,
	playlistId: RundownPlaylistId
): VerifiedRundownPlaylistContentAccess {
	const access = RundownPlaylistContentWriteAccess.playout(context, playlistId)
	const playlist = access.playlist
	if (!playlist) throw new Meteor.Error(404, `Rundown Playlist "${playlistId}" not found!`)
	return {
		...access,
		playlist,
	}
}

export function checkAccessAndGetPlaylist(context: MethodContext, playlistId: RundownPlaylistId): RundownPlaylist {
	const access = RundownPlaylistContentWriteAccess.playout(context, playlistId)
	const playlist = access.playlist
	if (!playlist) throw new Meteor.Error(404, `Rundown Playlist "${playlistId}" not found!`)
	return playlist
}

export function checkAccessAndGetRundown(context: MethodContext, rundownId: RundownId): Rundown {
	const access = RundownPlaylistContentWriteAccess.rundown(context, rundownId)
	const rundown = access.rundown
	if (!rundown) throw new Meteor.Error(404, `Rundown "${rundownId}" not found!`)
	return rundown
}
