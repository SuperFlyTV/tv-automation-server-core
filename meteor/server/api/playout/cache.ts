import { ReadonlyDeep } from 'type-fest'
import _ from 'underscore'
import { PartInstance, DBPartInstance, PartInstances } from '../../../lib/collections/PartInstances'
import { Part, DBPart, Parts } from '../../../lib/collections/Parts'
import { PeripheralDevice, PeripheralDevices } from '../../../lib/collections/PeripheralDevices'
import { PieceInstance, PieceInstances } from '../../../lib/collections/PieceInstances'
import {
	RundownPlaylist,
	DBRundownPlaylist,
	RundownPlaylistId,
	RundownPlaylists,
} from '../../../lib/collections/RundownPlaylists'
import { Rundown, DBRundown, Rundowns } from '../../../lib/collections/Rundowns'
import { Segment, DBSegment, Segments } from '../../../lib/collections/Segments'
import { Studio, StudioId, Studios } from '../../../lib/collections/Studios'
import { Timeline, TimelineComplete } from '../../../lib/collections/Timeline'
import { ActivationCache, getActivationCache } from '../../cache/ActivationCache'
import { DbCacheReadCollection, DbCacheWriteCollection } from '../../cache/CacheCollection'
import { DbCacheReadObject, DbCacheWriteObject } from '../../cache/CacheObject'
import { CacheBase, ReadOnlyCache } from '../../cache/CacheBase'
import { profiler } from '../profiler'
import { removeRundownPlaylistFromDb } from '../rundownPlaylist'
import { CacheForStudioBase } from '../studio/cache'
import { getRundownsSegmentsAndPartsFromCache } from './lib'
import { CacheForIngest } from '../ingest/cache'
import { Meteor } from 'meteor/meteor'

/**
 * This is a cache used for playout operations.
 * It is intentionally very lightweight, with the intention of it to be used only for some initial verification that a playout operation can be performed.
 */
export abstract class CacheForPlayoutPreInit extends CacheBase<CacheForPlayout> {
	public readonly isPlayout = true
	public readonly PlaylistId: RundownPlaylistId

	public readonly activationCache: ActivationCache

	public readonly Studio: DbCacheReadObject<Studio, Studio>
	public readonly PeripheralDevices: DbCacheReadCollection<PeripheralDevice, PeripheralDevice>

	public readonly Playlist: DbCacheWriteObject<RundownPlaylist, DBRundownPlaylist>
	public readonly Rundowns: DbCacheReadCollection<Rundown, DBRundown>

	protected constructor(studioId: StudioId, playlistId: RundownPlaylistId) {
		super()

		this.PlaylistId = playlistId
		this.activationCache = getActivationCache(studioId, playlistId)

		this.Studio = new DbCacheReadObject(Studios, false)
		this.PeripheralDevices = new DbCacheReadCollection(PeripheralDevices)

		this.Playlist = new DbCacheWriteObject(RundownPlaylists, false)
		this.Rundowns = new DbCacheReadCollection(Rundowns)
	}

	protected async preInit(tmpPlaylist: ReadonlyDeep<RundownPlaylist>) {
		await Promise.allSettled([
			this.Playlist._initialize(tmpPlaylist._id),
			this.Rundowns.prepareInit({ playlistId: tmpPlaylist._id }, true),
		])

		const rundowns = this.Rundowns.findFetch()
		await this.activationCache.initialize(this.Playlist.doc, rundowns)

		this.Studio._fromDoc(this.activationCache.getStudio())
		await this.PeripheralDevices.prepareInit(async () => {
			const data = await this.activationCache.getPeripheralDevices()
			this.PeripheralDevices.fillWithDataFromArray(data)
		}, true)
	}
}

/**
 * This is a cache used for playout operations.
 * It contains everything that is needed to generate the timeline, and everything except for pieces needed to update the partinstances.
 * Anything not in this cache should not be needed often, and only for specific operations (eg, AdlibActions needed to run one).
 */
export class CacheForPlayout extends CacheForPlayoutPreInit implements CacheForStudioBase {
	private toBeRemoved: boolean = false

	public readonly Timeline: DbCacheWriteCollection<TimelineComplete, TimelineComplete>

	public readonly Segments: DbCacheReadCollection<Segment, DBSegment>
	public readonly Parts: DbCacheReadCollection<Part, DBPart>
	public readonly PartInstances: DbCacheWriteCollection<PartInstance, DBPartInstance>
	public readonly PieceInstances: DbCacheWriteCollection<PieceInstance, PieceInstance>

	protected constructor(studioId: StudioId, playlistId: RundownPlaylistId) {
		super(studioId, playlistId)

		this.Timeline = new DbCacheWriteCollection<TimelineComplete, TimelineComplete>(Timeline)

		this.Segments = new DbCacheReadCollection<Segment, DBSegment>(Segments)
		this.Parts = new DbCacheReadCollection<Part, DBPart>(Parts)

		this.PartInstances = new DbCacheWriteCollection<PartInstance, DBPartInstance>(PartInstances)
		this.PieceInstances = new DbCacheWriteCollection<PieceInstance, PieceInstance>(PieceInstances)
	}

	static async create(tmpPlaylist: ReadonlyDeep<RundownPlaylist>): Promise<CacheForPlayout> {
		const res = new CacheForPlayout(tmpPlaylist.studioId, tmpPlaylist._id)

		await res.preInit(tmpPlaylist)

		return res
	}

	static async from(
		newPlaylist: ReadonlyDeep<RundownPlaylist>,
		newRundowns: ReadonlyDeep<Array<Rundown>>,
		ingestCache: ReadOnlyCache<CacheForIngest>
	): Promise<CacheForPlayout> {
		const res = new CacheForPlayout(newPlaylist.studioId, newPlaylist._id)

		res.Playlist._fromDoc(newPlaylist)
		await res.Rundowns.prepareInit(async () => {
			// newRundowns should already contain the update Rundown from ingestCache
			res.Rundowns.fillWithDataFromArray(newRundowns)
		}, true)

		await res.preInit(res.Playlist.doc)

		await res.initContent(ingestCache)

		return res
	}

	/**
	 * Intitialise the full content of the cache
	 * @param ingestCache A CacheForIngest that is pending saving, if this is following an ingest operation
	 */
	async initContent(ingestCache: ReadOnlyCache<CacheForIngest> | null): Promise<void> {
		const playlist = this.Playlist.doc

		const ps: Promise<any>[] = []

		const rundownIds = this.Rundowns.findFetch().map((r) => r._id)

		const selectedPartInstanceIds = _.compact([
			playlist.currentPartInstanceId,
			playlist.nextPartInstanceId,
			playlist.previousPartInstanceId,
		])

		// If there is an ingestCache, then avoid loading some bits from the db for that rundown
		const loadRundownIds = ingestCache ? rundownIds.filter((id) => id !== ingestCache.RundownId) : rundownIds
		ps.push(this.Segments.prepareInit({ rundownId: { $in: loadRundownIds } }, true))
		ps.push(this.Parts.prepareInit({ rundownId: { $in: loadRundownIds } }, true))

		ps.push(
			// Load all instances which are not reset, or which are selected
			this.PartInstances.prepareInit(
				{
					playlistActivationId: playlist.activationId,
					rundownId: { $in: rundownIds },
					$or: [
						{
							reset: { $ne: true },
						},
						{
							_id: { $in: selectedPartInstanceIds },
						},
					],
				},
				true
			)
		)

		ps.push(
			// Load all instances for the selected partInstances
			this.PieceInstances.prepareInit(
				{
					playlistActivationId: playlist.activationId,
					rundownId: { $in: rundownIds },
					partInstanceId: { $in: selectedPartInstanceIds },
				},
				true
			)
		)

		// Future: This could be defered until we get to updateTimeline. It could be a small performance boost
		ps.push(this.Timeline.prepareInit({ _id: playlist.studioId }, true))

		await Promise.allSettled(ps)

		if (ingestCache) {
			// Populate the collections with the cached data instead
			this.Segments.fillWithDataFromArray(ingestCache.Segments.findFetch())
			this.Parts.fillWithDataFromArray(ingestCache.Parts.findFetch())
		}
	}

	/**
	 * Remove the playlist when this cache is saved.
	 * The cache is cleared of any documents, and any deferred functions are discarded
	 * Note: any deferred functions that get added after this will be ignoted
	 */
	removePlaylist() {
		if (this.Playlist.doc.activationId) {
			throw new Meteor.Error(500, 'Cannot remove the active RundownPlaylist')
		}
		this.toBeRemoved = true

		super.markCollectionsForRemoval()
	}

	discardChanges() {
		this._abortActiveTimeout()

		this.toBeRemoved = false
		super.discardChanges()

		// Discard any hooks too
		this._deferredAfterSaveFunctions.length = 0
		this._deferredFunctions.length = 0

		this.assertNoChanges()
	}

	async saveAllToDatabase() {
		if (this.toBeRemoved) {
			const span = profiler.startSpan('CacheForPlayout.saveAllToDatabase')
			this._abortActiveTimeout()

			// Ignoring any deferred functions

			await removeRundownPlaylistFromDb(this.Playlist.doc)

			super.assertNoChanges()
			span?.end()
		} else {
			return super.saveAllToDatabase()
		}
	}
}

export function getOrderedSegmentsAndPartsFromPlayoutCache(
	cache: CacheForPlayout
): {
	segments: Segment[]
	parts: Part[]
} {
	const rundowns = cache.Rundowns.findFetch(
		{},
		{
			sort: {
				_rank: 1,
				_id: 1,
			},
		}
	)
	return getRundownsSegmentsAndPartsFromCache(cache.Parts, cache.Segments, rundowns)
}
export function getRundownIDsFromCache(cache: CacheForPlayout) {
	return cache.Rundowns.findFetch({}).map((r) => r._id)
}
export function getSelectedPartInstancesFromCache(
	cache: CacheForPlayout
): {
	currentPartInstance: PartInstance | undefined
	nextPartInstance: PartInstance | undefined
	previousPartInstance: PartInstance | undefined
} {
	const playlist = cache.Playlist.doc

	return {
		currentPartInstance: playlist.currentPartInstanceId
			? cache.PartInstances.findOne(playlist.currentPartInstanceId)
			: undefined,
		nextPartInstance: playlist.nextPartInstanceId
			? cache.PartInstances.findOne(playlist.nextPartInstanceId)
			: undefined,
		previousPartInstance: playlist.previousPartInstanceId
			? cache.PartInstances.findOne(playlist.previousPartInstanceId)
			: undefined,
	}
}
