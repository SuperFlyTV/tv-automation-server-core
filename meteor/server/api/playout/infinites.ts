import * as _ from 'underscore'
import { DBPart } from '../../../lib/collections/Parts'
import { Piece, Pieces } from '../../../lib/collections/Pieces'
import { PartInstance, PartInstanceId } from '../../../lib/collections/PartInstances'
import { PieceInstance } from '../../../lib/collections/PieceInstances'
import { RundownPlaylist } from '../../../lib/collections/RundownPlaylists'
import { PartsAndSegments, selectNextPart } from './lib'
import { saveIntoCache } from '../../cache/lib'
import {
	getPieceInstancesForPart as libgetPieceInstancesForPart,
	getPlayheadTrackingInfinitesForPart as libgetPlayheadTrackingInfinitesForPart,
	buildPiecesStartingInThisPartQuery,
	buildPastInfinitePiecesForThisPartQuery,
	processAndPrunePieceInstanceTimings,
} from '../../../lib/rundown/infinites'
import { profiler } from '../profiler'
import { Meteor } from 'meteor/meteor'
import { CacheForPlayout, getOrderedSegmentsAndPartsFromPlayoutCache, getSelectedPartInstancesFromCache } from './cache'
import { ReadonlyDeep } from 'type-fest'
import { asyncCollectionFindFetch } from '../../lib/database'
import { getCurrentTime } from '../../../lib/lib'
import { CacheForIngest } from '../ingest/cache'
import { ReadOnlyCache } from '../../cache/CacheBase'

// /** When we crop a piece, set the piece as "it has definitely ended" this far into the future. */
export const DEFINITELY_ENDED_FUTURE_DURATION = 1 * 1000

/**
 * We can only continue adlib onEnd infinites if we go forwards in the rundown. Any distance backwards will clear them.
 * */
function canContinueAdlibOnEndInfinites(
	playlist: ReadonlyDeep<RundownPlaylist>,
	orderedPartsAndSegments: PartsAndSegments,
	previousPartInstance: PartInstance | undefined,
	part: DBPart
): boolean {
	if (previousPartInstance && playlist) {
		const span = profiler.startSpan('canContinueAdlibOnEndInfinites')
		// TODO - if we don't have an index for previousPartInstance, what should we do?

		const expectedNextPart = selectNextPart(playlist, previousPartInstance, orderedPartsAndSegments)
		if (expectedNextPart) {
			if (expectedNextPart.part._id === part._id) {
				// Next part is what we expect, so take it
				return true
			} else {
				const partIndex = orderedPartsAndSegments.parts.findIndex((p) => p._id === part._id)
				if (partIndex >= expectedNextPart.index) {
					if (span) span.end()
					// Somewhere after the auto-next part, so we can use that
					return true
				} else {
					if (span) span.end()
					// It isnt ahead, so we cant take it
					return false
				}
			}
		} else {
			if (span) span.end()
			// selectNextPart gave nothing, so we must be at the end?
			return false
		}
	} else {
		// There won't be anything to continue anyway..
		return false
	}
}

function getIdsBeforeThisPart(cache: CacheForPlayout, nextPart: DBPart) {
	const span = profiler.startSpan('getIdsBeforeThisPart')
	// Get the normal parts
	const partsBeforeThisInSegment = cache.Parts.findFetch(
		(p) => p.segmentId === nextPart.segmentId && p._rank < nextPart._rank
	)
	// Find any orphaned parts
	const partInstancesBeforeThisInSegment = cache.PartInstances.findFetch(
		(p) => p.segmentId === nextPart.segmentId && p.orphaned && p.part._rank < nextPart._rank
	)
	partsBeforeThisInSegment.push(...partInstancesBeforeThisInSegment.map((p) => p.part))

	const currentSegment = cache.Segments.findOne(nextPart.segmentId)
	const segmentsBeforeThisInRundown = currentSegment
		? cache.Segments.findFetch({
				rundownId: nextPart.rundownId,
				_rank: { $lt: currentSegment._rank },
		  }).map((p) => p._id)
		: []

	if (span) span.end()
	return {
		partsBeforeThisInSegment: _.sortBy(partsBeforeThisInSegment, (p) => p._rank).map((p) => p._id),
		segmentsBeforeThisInRundown,
	}
}

export async function fetchPiecesThatMayBeActiveForPart(
	cache: CacheForPlayout,
	unsavedIngestCache: Omit<ReadOnlyCache<CacheForIngest>, 'Rundown'> | undefined,
	part: DBPart
): Promise<Piece[]> {
	const span = profiler.startSpan('fetchPiecesThatMayBeActiveForPart')

	const thisPiecesQuery = buildPiecesStartingInThisPartQuery(part)
	const pPiecesStartingInPart = unsavedIngestCache
		? Promise.resolve(unsavedIngestCache.Pieces.findFetch(thisPiecesQuery))
		: asyncCollectionFindFetch(Pieces, thisPiecesQuery)

	const { partsBeforeThisInSegment, segmentsBeforeThisInRundown } = getIdsBeforeThisPart(cache, part)

	const infinitePiecesQuery = buildPastInfinitePiecesForThisPartQuery(
		part,
		partsBeforeThisInSegment,
		segmentsBeforeThisInRundown
	)
	// Future scope: Once there is a longer than Rundown infinite mode, this will need to split the query to search everywhere
	const pInfinitePieces = unsavedIngestCache
		? Promise.resolve(unsavedIngestCache.Pieces.findFetch(infinitePiecesQuery))
		: asyncCollectionFindFetch(Pieces, infinitePiecesQuery)

	const [piecesStartingInPart, infinitePieces] = await Promise.all([pPiecesStartingInPart, pInfinitePieces])
	if (span) span.end()
	return [...piecesStartingInPart, ...infinitePieces]
}

export async function syncPlayheadInfinitesForNextPartInstance(cache: CacheForPlayout): Promise<void> {
	const span = profiler.startSpan('syncPlayheadInfinitesForNextPartInstance')
	const { nextPartInstance, currentPartInstance } = getSelectedPartInstancesFromCache(cache)
	if (nextPartInstance && currentPartInstance) {
		const playlist = cache.Playlist.doc
		if (!playlist.activationId) throw new Meteor.Error(500, `RundownPlaylist "${playlist._id}" is not active`)

		const { partsBeforeThisInSegment, segmentsBeforeThisInRundown } = getIdsBeforeThisPart(
			cache,
			nextPartInstance.part
		)

		const rundown = cache.Rundowns.findOne(currentPartInstance.rundownId)
		if (!rundown) throw new Meteor.Error(404, `Rundown "${currentPartInstance.rundownId}" not found!`)

		// !! Database call !!
		const showStyleBase = await cache.activationCache.getShowStyleBase(rundown)

		const orderedPartsAndSegments = getOrderedSegmentsAndPartsFromPlayoutCache(cache)

		const canContinueAdlibOnEnds = canContinueAdlibOnEndInfinites(
			playlist,
			orderedPartsAndSegments,
			currentPartInstance,
			nextPartInstance.part
		)
		const playingPieceInstances = cache.PieceInstances.findFetch(
			(p) => p.partInstanceId === currentPartInstance._id
		)

		const nowInPart = getCurrentTime() - (currentPartInstance.timings?.startedPlayback ?? 0)
		const prunedPieceInstances = processAndPrunePieceInstanceTimings(
			showStyleBase,
			playingPieceInstances,
			nowInPart,
			undefined,
			true
		)

		const infinites = libgetPlayheadTrackingInfinitesForPart(
			playlist.activationId,
			new Set(partsBeforeThisInSegment),
			new Set(segmentsBeforeThisInRundown),
			currentPartInstance,
			prunedPieceInstances,
			nextPartInstance.part,
			nextPartInstance._id,
			canContinueAdlibOnEnds,
			false
		)

		saveIntoCache(
			cache.PieceInstances,
			{
				partInstanceId: nextPartInstance._id,
				'infinite.fromPreviousPlayhead': true,
			},
			infinites
		)
	}
	if (span) span.end()
}

export function getPieceInstancesForPart(
	cache: CacheForPlayout,
	playingPartInstance: PartInstance | undefined,
	part: DBPart,
	possiblePieces: Piece[],
	newInstanceId: PartInstanceId,
	isTemporary: boolean
): PieceInstance[] {
	const span = profiler.startSpan('getPieceInstancesForPart')
	const { partsBeforeThisInSegment, segmentsBeforeThisInRundown } = getIdsBeforeThisPart(cache, part)

	const playlist = cache.Playlist.doc
	if (!playlist.activationId) throw new Meteor.Error(500, `RundownPlaylist "${playlist._id}" is not active`)

	const orderedPartsAndSegments = getOrderedSegmentsAndPartsFromPlayoutCache(cache)
	const playingPieceInstances = playingPartInstance
		? cache.PieceInstances.findFetch((p) => p.partInstanceId === playingPartInstance._id)
		: []

	const canContinueAdlibOnEnds = canContinueAdlibOnEndInfinites(
		playlist,
		orderedPartsAndSegments,
		playingPartInstance,
		part
	)

	const res = libgetPieceInstancesForPart(
		playlist.activationId,
		playingPartInstance,
		playingPieceInstances,
		part,
		new Set(partsBeforeThisInSegment),
		new Set(segmentsBeforeThisInRundown),
		possiblePieces,
		orderedPartsAndSegments.parts.map((p) => p._id),
		newInstanceId,
		canContinueAdlibOnEnds,
		isTemporary
	)
	if (span) span.end()
	return res
}
