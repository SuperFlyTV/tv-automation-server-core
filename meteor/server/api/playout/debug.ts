import { Meteor } from 'meteor/meteor'
import { logger } from '../../logging'
import { waitForPromise, waitForPromiseAll } from '../../../lib/lib'
import { RundownPlaylists, RundownPlaylistId } from '../../../lib/collections/RundownPlaylists'
import { Settings } from '../../../lib/Settings'
import { setNextPart } from './lib'
import { syncPlayheadInfinitesForNextPartInstance } from './infinites'
import { forceClearAllActivationCaches } from '../../cache/ActivationCache'
import { PartInstances } from '../../../lib/collections/PartInstances'
import { PieceInstances } from '../../../lib/collections/PieceInstances'
import { updateStudioOrPlaylistTimeline, updateTimeline } from './timeline'
import { forceClearAllBlueprintConfigCaches } from '../blueprints/config'
import {
	PlayoutLockFunctionPriority,
	runPlayoutOperationWithCache,
	runPlayoutOperationWithCacheFromStudioOperation,
} from './lockFunction'
import { getSelectedPartInstancesFromCache } from './cache'
import { removeRundownPlaylistFromDb } from '../rundownPlaylist'
import { check } from 'meteor/check'
import { StudioId } from '../../../lib/collections/Studios'
import { ensureNextPartIsValid } from '../ingest/updateNext'
import { runStudioOperationWithCache, StudioLockFunctionPriority } from '../studio/lockFunction'

if (!Settings.enableUserAccounts) {
	// These are temporary method to fill the rundown database with some sample data
	// for development

	Meteor.methods({
		debug_removePlaylist(id: RundownPlaylistId) {
			logger.debug('Remove rundown "' + id + '"')

			const playlist = RundownPlaylists.findOne(id)
			if (!playlist) throw new Meteor.Error(404, `RundownPlaylist "${id}" not found`)
			waitForPromise(removeRundownPlaylistFromDb(playlist))
		},

		debug_removeAllPlaylists() {
			logger.debug('Remove all rundowns')

			waitForPromiseAll(RundownPlaylists.find({}).map((playlist) => removeRundownPlaylistFromDb(playlist)))
		},

		debug_updateTimeline: (studioId: StudioId) => {
			try {
				check(studioId, String)
				logger.info(`debug_updateTimeline: "${studioId}"`)

				runStudioOperationWithCache(
					'debug_updateTimeline',
					studioId,
					StudioLockFunctionPriority.USER_PLAYOUT,
					(cache) => {
						updateStudioOrPlaylistTimeline(cache)
					}
				)
			} catch (e) {
				logger.error(e)
				throw e
			}
		},
		debug_updateNext: (studioId: StudioId) => {
			try {
				check(studioId, String)
				logger.info(`debug_updateNext: "${studioId}"`)

				runStudioOperationWithCache(
					'debug_updateNext',
					studioId,
					StudioLockFunctionPriority.USER_PLAYOUT,
					(cache) => {
						const playlists = cache.getActiveRundownPlaylists()
						if (playlists.length === 1) {
							return runPlayoutOperationWithCacheFromStudioOperation(
								'updateStudioOrPlaylistTimeline',
								cache,
								playlists[0],
								PlayoutLockFunctionPriority.USER_PLAYOUT,
								null,
								(playlistCache) => {
									ensureNextPartIsValid(playlistCache)
								}
							)
						} else {
							throw new Error('No playlist active')
						}
					}
				)
			} catch (e) {
				logger.error(e)
				throw e
			}
		},

		debug_syncPlayheadInfinitesForNextPartInstance(id: RundownPlaylistId) {
			logger.info(`syncPlayheadInfinitesForNextPartInstance ${id}`)

			runPlayoutOperationWithCache(
				null,
				'debug_syncPlayheadInfinitesForNextPartInstance',
				id,
				PlayoutLockFunctionPriority.MISC,
				null,
				async (cache) => {
					await syncPlayheadInfinitesForNextPartInstance(cache)
				}
			)
		},

		debug_forceClearAllCaches() {
			logger.info('forceClearAllCaches')

			forceClearAllActivationCaches()
			forceClearAllBlueprintConfigCaches()
		},

		debug_clearAllResetInstances() {
			logger.info('clearAllResetInstances')

			PartInstances.remove({ reset: true })
			PieceInstances.remove({ reset: true })
		},

		debug_regenerateNextPartInstance(id: RundownPlaylistId) {
			logger.info('regenerateNextPartInstance')

			runPlayoutOperationWithCache(
				null,
				'debug_regenerateNextPartInstance',
				id,
				PlayoutLockFunctionPriority.MISC,
				null,
				(cache) => {
					const playlist = cache.Playlist.doc
					if (playlist.nextPartInstanceId && playlist.activationId) {
						const { nextPartInstance } = getSelectedPartInstancesFromCache(cache)
						const part = nextPartInstance ? cache.Parts.findOne(nextPartInstance.part._id) : undefined
						if (part) {
							setNextPart(cache, null)
							setNextPart(cache, { part: part })

							updateTimeline(cache)
						}
					}
				}
			)
		},
	})
}
