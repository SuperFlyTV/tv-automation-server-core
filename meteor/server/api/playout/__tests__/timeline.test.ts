import { Meteor } from 'meteor/meteor'
import '../../../../__mocks__/_extendJest'
import { testInFiber } from '../../../../__mocks__/helpers/jest'
import { fixSnapshot } from '../../../../__mocks__/helpers/snapshot'
import { setupDefaultStudioEnvironment, DefaultEnvironment, setupDefaultRundown } from '../../../../__mocks__/helpers/database'
import { Rundowns, Rundown } from '../../../../lib/collections/Rundowns'
import '../api'
import { Timeline } from '../../../../lib/collections/Timeline'
import { ServerPlayoutAPI } from '../playout'
import { updateTimeline } from '../timeline'

describe('Timeline', () => {
	let env: DefaultEnvironment
	beforeEach(() => {
		env = setupDefaultStudioEnvironment()
	})
	testInFiber('non-existing studio', () => {
		expect(() => {
			updateTimeline('asdf')
		}).toThrowError(/not found/i)
	})
	testInFiber('Basic rundown', () => {
		const rundownId0 = setupDefaultRundown(env)
		expect(rundownId0).toBeTruthy()

		const getRundown0 = () => {
			return Rundowns.findOne(rundownId0) as Rundown
		}
		const parts = getRundown0().getParts()

		expect(getRundown0()).toMatchObject({
			active: false,
			rehearsal: false
		})

		// Prepare and activate in rehersal:
		ServerPlayoutAPI.activateRundown(rundownId0, false)
		expect(getRundown0()).toMatchObject({
			active: true,
			rehearsal: false,
			currentPartId: null,
			nextPartId: parts[0]._id,
		})

		// Take the first Part:
		ServerPlayoutAPI.takeNextPart(rundownId0)
		expect(getRundown0()).toMatchObject({
			currentPartId: parts[0]._id,
			nextPartId: parts[1]._id,
		})

		updateTimeline(getRundown0().studioId)

		expect(fixSnapshot(Timeline.find().fetch())).toMatchSnapshot()

		const currentTime = 100 * 1000
		updateTimeline(getRundown0().studioId, currentTime)

		expect(fixSnapshot(Timeline.find().fetch())).toMatchSnapshot()

		// Deactivate rundown:
		ServerPlayoutAPI.deactivateRundown(rundownId0)
		expect(getRundown0()).toMatchObject({
			active: false,
			currentPartId: null,
			nextPartId: null
		})

		expect(fixSnapshot(Timeline.find().fetch())).toMatchSnapshot()
	})
})
