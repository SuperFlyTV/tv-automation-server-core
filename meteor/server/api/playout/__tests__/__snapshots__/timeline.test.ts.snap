// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timeline Basic rundown 1`] = `
Array [
  Object {
    "_id": "mockStudio10_randomId11002_status",
    "classes": Array [
      "rundown_active",
    ],
    "content": Object {
      "deviceType": 0,
    },
    "enable": Object {
      "while": 1,
    },
    "id": "randomId11002_status",
    "layer": "rundown_status",
    "objectType": "rundown",
    "rundownId": "randomId11002",
    "studioId": "mockStudio10",
  },
  Object {
    "_id": "mockStudio10_part_group_randomId11002_part0_0",
    "children": Array [],
    "content": Object {
      "deviceType": 0,
      "type": "group",
    },
    "enable": Object {
      "start": "now",
    },
    "id": "part_group_randomId11002_part0_0",
    "isGroup": true,
    "isPartGroup": true,
    "layer": "",
    "objectType": "rundown",
    "priority": 5,
    "rundownId": "randomId11002",
    "studioId": "mockStudio10",
  },
  Object {
    "_id": "mockStudio10_part_group_firstobject_randomId11002_part0_0",
    "classes": Array [],
    "content": Object {
      "callBack": "partPlaybackStarted",
      "callBackData": Object {
        "partId": "randomId11002_part0_0",
        "rundownId": "randomId11002",
      },
      "callBackStopped": "partPlaybackStopped",
      "deviceType": 0,
      "type": "callback",
    },
    "enable": Object {
      "start": 0,
    },
    "id": "part_group_firstobject_randomId11002_part0_0",
    "inGroup": "part_group_randomId11002_part0_0",
    "layer": "group_first_object",
    "objectType": "rundown",
    "rundownId": "randomId11002",
    "studioId": "mockStudio10",
  },
  Object {
    "_id": "mockStudio10_statObj",
    "content": Object {
      "deviceType": 0,
      "objCount": 3,
      "type": "nothing",
    },
    "enable": Object {
      "start": 0,
    },
    "id": "statObj",
    "layer": "__stat",
    "objectType": "stat",
    "studioId": "mockStudio10",
  },
]
`;

exports[`Timeline Basic rundown 2`] = `
Array [
  Object {
    "_id": "mockStudio10_randomId11002_status",
    "classes": Array [
      "rundown_active",
    ],
    "content": Object {
      "deviceType": 0,
    },
    "enable": Object {
      "while": 1,
    },
    "id": "randomId11002_status",
    "layer": "rundown_status",
    "objectType": "rundown",
    "rundownId": "randomId11002",
    "studioId": "mockStudio10",
  },
  Object {
    "_id": "mockStudio10_part_group_firstobject_randomId11002_part0_0",
    "classes": Array [],
    "content": Object {
      "callBack": "partPlaybackStarted",
      "callBackData": Object {
        "partId": "randomId11002_part0_0",
        "rundownId": "randomId11002",
      },
      "callBackStopped": "partPlaybackStopped",
      "deviceType": 0,
      "type": "callback",
    },
    "enable": Object {
      "start": 0,
    },
    "id": "part_group_firstobject_randomId11002_part0_0",
    "inGroup": "part_group_randomId11002_part0_0",
    "layer": "group_first_object",
    "objectType": "rundown",
    "rundownId": "randomId11002",
    "studioId": "mockStudio10",
  },
  Object {
    "_id": "mockStudio10_part_group_randomId11002_part0_0",
    "children": Array [],
    "content": Object {
      "deviceType": 0,
      "type": "group",
    },
    "enable": Object {
      "setFromNow": true,
      "start": 100000,
    },
    "id": "part_group_randomId11002_part0_0",
    "isGroup": true,
    "isPartGroup": true,
    "layer": "",
    "objectType": "rundown",
    "priority": 5,
    "rundownId": "randomId11002",
    "studioId": "mockStudio10",
  },
  Object {
    "_id": "mockStudio10_statObj",
    "content": Object {
      "deviceType": 0,
      "objCount": 3,
      "type": "nothing",
    },
    "enable": Object {
      "start": 0,
    },
    "id": "statObj",
    "layer": "__stat",
    "objectType": "stat",
    "studioId": "mockStudio10",
  },
]
`;

exports[`Timeline Basic rundown 3`] = `
Array [
  Object {
    "_id": "mockStudio10_statObj",
    "content": Object {
      "deviceType": 0,
      "objCount": 1,
      "type": "nothing",
    },
    "enable": Object {
      "start": 0,
    },
    "id": "statObj",
    "layer": "__stat",
    "objectType": "stat",
    "studioId": "mockStudio10",
  },
  Object {
    "_id": "mockStudio10_baseline_version",
    "content": Object {
      "deviceType": 0,
    },
    "enable": Object {
      "start": 0,
    },
    "id": "baseline_version",
    "layer": "baseline_version",
    "metadata": Object {
      "versions": Object {
        "blueprintId": "mockBlueprint6",
        "blueprintVersion": "0.0.0",
        "core": "0.0.0-test",
        "studio": "asdf",
      },
    },
    "objectType": "rundown",
    "rundownId": "",
    "studioId": "mockStudio10",
  },
]
`;
